lib/good_result.rb

```diff
@@ -3,6 +3,45 @@
 require_relative "good_result/version"
 
 module GoodResult
-  class Error < StandardError; end
-  # Your code goes here...
+  class ResultError < StandardError; end
+  
+  class Success
+    attr_reader :value
+
+    def initialize(value)
+      @value = value
+    end
+
+    def success?
+      true
+    end
+
+    def failure?
+      false
+    end
+
+    def error
+      raise ResultError, 'Expected error but Result is Success'
+    end
```

Comment:
My only question here is more of an API one:

- Do we want this `error` method to "blow up" when someone calls it, without a more explicit callout? ie. Naming the method `error!` or `throw_error` etc.
- Do we want this `error` method to be `public` at all. Given the wording of the error message itself, my biggest concern is that the enforcement of that seems to becoming from _outside_ the class itself. Meaning, we have to _trust_ whoever used this monad used it properly. Instead of being able to trust in the Result object itself.

Reply:
It's just kind of in case the caller tries to access a value that wouldn't exist on that object. So if the caller tries to access an error on a Success or a value on a Failure.

I could alternatively just return `nil` on those instead of raising an error

Reply:
Ah okay, that makes sense! From what you're saying, it sounds like this `error` method should be private, so no-one calls it externally on accident.

Reply:
`Success` would never really need to call `error` privately within the class, I was including it more as defensive coding, in case the caller thinks they have a `Failure` and tries to access the `error`, if they do that on a `Success` they would get a `NoMethodError`.

Granted, a caller would probably call `result.success?` to check the result, so this was more of a edge case / just in case sort of thing 😅

I got rid of it though, as it is probably an edge case.